paasta_tools/paasta_serviceinit.py

```diff
@@ -143,57 +145,107 @@ def main():
-            elif instance_type == 'paasta_native':
-                return_code = paasta_native_serviceinit.perform_command(
-                    command=command,
-                    service=service,
-                    instance=instance,
-                    cluster=cluster,
-                    verbose=args.verbose,
-                    soa_dir=args.soa_dir,
+            )
+            return_codes.append(1)
+        else:
+            instance_types_map[instance_type].append(instance)
+
+
+    remote_run_frameworks = None
+    if len(instance_types_map['adhoc']) > 0:
+        remote_run_frameworks = paasta_remote_run.remote_run_frameworks()
+
+    for instance_type in instance_types:
+        for instance in instance_types_map[instance_type]:
+            try:
+                version = get_deployment_version(
+                    actual_deployments, cluster, instance
                 )
-            else:
-                log.error("I calculated an instance_type of %s for %s which I don't know how to handle."
-                          % (instance_type, compose_job_id(service, instance)))
+                paasta_print('instance: %s' % PaastaColors.blue(instance))
+                paasta_print('Git sha:    %s (desired)' % version)
+
+                if instance_type == 'marathon':
+                    return_code = marathon_serviceinit.perform_command(
+                        command=command,
+                        service=service,
+                        instance=instance,
+                        cluster=cluster,
+                        verbose=args.verbose,
+                        soa_dir=args.soa_dir,
+                        app_id=args.app_id,
+                        delta=args.delta,
+                        client=clients.marathon(),
+                    )
+                elif instance_type == 'chronos':
+                    return_code = chronos_serviceinit.perform_command(
+                        command=command,
+                        service=service,
+                        instance=instance,
+                        cluster=cluster,
+                        verbose=args.verbose,
+                        soa_dir=args.soa_dir,
+                        client=clients.chronos(),
+                    )
+                elif instance_type == 'paasta_native':
+                    return_code = paasta_native_serviceinit.perform_command(
+                        command=command,
+                        service=service,
+                        instance=instance,
+                        cluster=cluster,
+                        verbose=args.verbose,
+                        soa_dir=args.soa_dir,
+                    )
+                elif instance_type == 'adhoc':
+                    if instance == 'interactive':
+                        continue
+                    if command != 'status':
+                        continue
+
+                    paasta_remote_run.remote_run_list(
+                        Dict(
```

Comment:
I'm actually kind of against spreading addict any further than it needs to reach.